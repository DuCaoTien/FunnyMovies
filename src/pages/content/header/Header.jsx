import { useState, useCallback, useEffect } from 'react';
import { useNavigate } from "react-router-dom";
import { toast } from 'react-toastify';
import isEmpty from 'lodash/isEmpty'
import bcrypt from 'bcryptjs'

import { isValidEmail } from "../../../helpers/helpers";

import 'react-toastify/dist/ReactToastify.css';
import './styles.scss';

const LOGIN_BTN_NAME = "btn_login";

function Header() {

    const [state, setState] = useState({
        email: "",
        password: "",
        isSignedIn: false
    });

    const { email, password, isSignedIn } = state;
    const navigate = useNavigate();

    useEffect(function () {
        const signedUser = JSON.parse(localStorage.getItem('signedUser')) || '';
        if (signedUser) {
            return setState((prevState) => {
                return {
                    ...prevState,
                    isSignedIn: true,
                    email: signedUser
                }
            })
        }
    }, [])

    const handleChange = useCallback((event) => {
        setState((prevState) => {
            return {
                ...prevState,
                [event.target.name]: event.target.value,
            }
        })
    }, []);

    const handleSubmit = useCallback(async (event) => {
        if (!isValidEmail(email)) {
            return toast.error("Email is invalid");
        }

        if (password.length < 6) {
            return toast.error("Password requires 6 characters minimum");
        }

        const accounts = JSON.parse(localStorage.getItem('accounts')) || [];
        const hashedPassword = bcrypt.hashSync(password);
        const duplicateEmailAccount = accounts.find((account) => account.email === email);

        if (event.target.name === LOGIN_BTN_NAME) {
            if (!isEmpty(duplicateEmailAccount)) {
                bcrypt.compare(password, duplicateEmailAccount.password, (err, isExistPassword) => {

                    if (err) {
                        return console.error(err);
                    }

                    if (isExistPassword) {
                        toast.success("Login successful");
                        localStorage.setItem('signedUser', JSON.stringify(email));
                        return setState((prevState) => {
                            return {
                                ...prevState,
                                isSignedIn: true
                            }
                        })
                    } else {
                        toast.error("Password does not match");
                    }

                });
            } else {
                return toast.error("User does not exist");
            }
        } else {
            if (!isEmpty(duplicateEmailAccount)) {
                return toast.error("Duplicate email");
            }

            const data = {
                email,
                password: hashedPassword
            };

            localStorage.setItem('accounts', JSON.stringify([...accounts, data]));
            localStorage.setItem('signedUser', JSON.stringify(email));
            setState((prevState) => {
                return {
                    ...prevState,
                    isSignedIn: true
                }
            })
            return toast.success("Register successful");
        }

    }, [email, password]);

    const handleLogout = useCallback(() => {
        localStorage.removeItem('signedUser');
        setState((prevState) => {
            return {
                ...prevState,
                isSignedIn: false
            }
        });
        if (typeof navigate === 'function') {
            navigate('/FunnyMovies/');
        }
    }, [navigate]);

    const handleShareMovie = useCallback(() => {
        if (typeof navigate === 'function') {
            navigate('/FunnyMovies/share');
        }
    }, [navigate]);

    const goHomePage = useCallback(() => {
        if (typeof navigate === 'function') {
            navigate('/FunnyMovies/');
        }
    }, [navigate]);

    return (
        <header>
            <div className="header_content">
                <div className="header_side">
                    <div className="logo" onClick={goHomePage}>FUNNY MOVIES</div>
                </div>
                <div className="header_side">
                    <div className='signin'>
                        {(!isSignedIn &&
                            <>
                                <div>
                                    <input
                                        id="email"
                                        name="email"
                                        className="input"
                                        type="text"
                                        value={email}
                                        onChange={handleChange}
                                        autoComplete="email"
                                        placeholder='Email'
                                    />
                                </div>
                                <div>
                                    <input
                                        id="password"
                                        name="password"
                                        className="input"
                                        type='password'
                                        value={password}
                                        autoComplete="new-password"
                                        onChange={handleChange}
                                        placeholder='Password'
                                    />
                                </div>
                                <div className="submit">
                                    <button
                                        className='btn'
                                        tabIndex="0"
                                        name={LOGIN_BTN_NAME}
                                        aria-label='login'
                                        onClick={handleSubmit}
                                        disabled={!email || !password}
                                    >
                                        Login
                                    </button>
                                </div>

                                <div className="submit">
                                    <button
                                        className='btn'
                                        tabIndex="0"
                                        name="btn-register"
                                        aria-label='register'
                                        onClick={handleSubmit}
                                        disabled={!email || !password}
                                    >
                                         Register
                                    </button>
                                </div>
                            </>)
                        ||
                        <>
                            <div className="">
                                <span>{`Welcome ${email}`}</span>
                            </div>
                            <div className="submit">
                                <button
                                    className='btn'
                                    tabIndex="0"
                                    name="btn_share"
                                    aria-label='share'
                                    onClick={handleShareMovie}
                                >
                                    Share a movie
                                </button>
                            </div>
                            <div className="submit">
                                <button
                                    className='btn'
                                    tabIndex="0"
                                    name="btn_logout"
                                    aria-label='logout'
                                    onClick={handleLogout}
                                >
                                    Log out
                                </button>
                            </div>
                        </>
                        }
                    </div>
                </div>
            </div>
        </header>
    );
}

export default Header;

