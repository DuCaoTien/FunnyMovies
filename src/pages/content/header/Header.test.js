import Header from "./Header";
import { render } from '@testing-library/react';

const mockedNavigator = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedNavigator,
    useLocation: jest.fn()
}));

test('Header runs as expected',  () => {
    const rendered = render(<Header />);
    expect(rendered).toMatchSnapshot();

    const { getByText } = rendered;

    expect(getByText('FUNNY MOVIES')).toBeInTheDocument();
    expect(getByText('Login')).toBeInTheDocument();
    expect(getByText('Register')).toBeInTheDocument();
})
