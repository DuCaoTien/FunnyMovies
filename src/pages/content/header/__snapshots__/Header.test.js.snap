// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header runs as expected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <div
          class="header_content"
        >
          <div
            class="header_side"
          >
            <div
              class="logo"
            >
              FUNNY MOVIES
            </div>
          </div>
          <div
            class="header_side"
          >
            <div
              class="signin"
            >
              <div>
                <input
                  autocomplete="email"
                  class="input"
                  id="email"
                  name="email"
                  placeholder="Email"
                  type="text"
                  value=""
                />
              </div>
              <div>
                <input
                  autocomplete="new-password"
                  class="input"
                  id="password"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="submit"
              >
                <button
                  aria-label="login"
                  class="btn"
                  disabled=""
                  name="btn_login"
                  tabindex="0"
                >
                  Login
                </button>
              </div>
              <div
                class="submit"
              >
                <button
                  aria-label="register"
                  class="btn"
                  disabled=""
                  name="btn-register"
                  tabindex="0"
                >
                  Register
                </button>
              </div>
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header>
      <div
        class="header_content"
      >
        <div
          class="header_side"
        >
          <div
            class="logo"
          >
            FUNNY MOVIES
          </div>
        </div>
        <div
          class="header_side"
        >
          <div
            class="signin"
          >
            <div>
              <input
                autocomplete="email"
                class="input"
                id="email"
                name="email"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div>
              <input
                autocomplete="new-password"
                class="input"
                id="password"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="submit"
            >
              <button
                aria-label="login"
                class="btn"
                disabled=""
                name="btn_login"
                tabindex="0"
              >
                Login
              </button>
            </div>
            <div
              class="submit"
            >
              <button
                aria-label="register"
                class="btn"
                disabled=""
                name="btn-register"
                tabindex="0"
              >
                Register
              </button>
            </div>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
